tests/fail/stacked_borrows/illegal_read_despite_exposed1.rs

```diff
@@ -0,0 +1,17 @@
+// compile-flags: -Zmiri-permissive-provenance
+
+fn main() {
+    unsafe {
+        let root = &mut 42;
+        let addr = root as *mut i32 as usize;
+        let exposed_ptr = addr as *mut i32;
+        // From the exposed ptr, we get a new unique ptr.
+        let root2 = &mut *exposed_ptr;
+        let _fool = root2 as *mut _; // this would have fooled the old untagged pointer logic
+        // Stack: Unknown(<N), Unique(N), SRW(N+1)
+        // And we test that it has uniqueness by doing a conflicting write.
+        *exposed_ptr = 0;
+        // Stack: Unknown(<N)
+        let _val = *root2; //~ ERROR: borrow stack
```

Comment:
This collapsing behavior seems very important and is worth having a note for, like we do for tag invalidation. That's certainly something I can do after this PR.

Reply:
Note that collapsing is *not* part of the spec -- it is part of the approximation Miri does because the exact spec (as documented [here](https://doc.rust-lang.org/nightly/std/ptr/fn.from_exposed_addr.html)) cannot be reasonably implemented. So, in the future, if we find a better way to approximate this, we might change this.

I guess what I am saying is -- this might be even harder to communicate than the regular Stacked Borrows rules, since these are derived rules that are computed from the actual rules.